backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/EnumWhenLowering.kt

```diff
@@ -143,3 +87,46 @@ internal class EnumWhenLowering(
                 IrDeclarationOrigin.IR_TEMPORARY_VARIABLE, ordinalDescriptor, getOrdinal)
     }
 }
+
+/**
+ * Replaces comparison of [subject] with enum entries with comparison of
+ * [subjectOrdinalProvider] and entry's ordinal.
+ */
+private class EnumEqEqTransformer(val context: Context, val subject: IrVariable, val subjectOrdinalProvider: Lazy<IrVariable>)
+    : IrElementTransformerVoid() {
+
+    override fun visitCall(expression: IrCall): IrExpression {
+        expression.transformChildrenVoid(this)
+        return tryLower(expression)
+    }
+
+    private val areEqualByValue = context.ir.symbols.areEqualByValue.first {
+        it.owner.valueParameters[0].type == context.builtIns.intType
+    }
+
+    // We are looking for branch that is a comparison of the subject and another enum entry.
+    // Both entries should belong to the same class.
+    fun tryLower(call: IrCall): IrExpression {
+        if (call.origin != IrStatementOrigin.EQEQ) {
+            return call
+        }
+        val callArgs = call.getArguments()
+        if (callArgs.size != 2) {
+            return call
+        }
+        val lhs = callArgs[0].second
+        val rhs = callArgs[1].second
+        if (lhs.type != rhs.type) {
+            return call
```

Comment:
What this check is required for?

Reply:
Without this check first branch will trigger in this case:
```kotlin
fun produceEntry() = EnumA.A

// Check that we fail on comparison of different enum types.
fun differentEnums() {
    println(when (produceEntry()) {
        EnumB.A -> "EnumB.A"
        EnumA.A -> "EnumA.A"
        EnumA.B -> "EnumA.B"
        else    -> "nah"
    })
}
```